drivers/timer/mchp_xec_rtos_timer.c

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <drivers/system_timer.h>
+#include <sys_clock.h>
+#include <spinlock.h>
+#include <arch/arm/cortex_m/cmsis.h>
+
+#define DEBUG_RTOS_TIMER
+
+#define COUNTER_MAX 0x0fffffffUL
+#define TIMER_STOPPED 0xf0000000UL
+
+#if 0
+#define CYC_PER_TICK (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC	\
+		      / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#else
+
+#define CYC_PER_TICK_M ((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC * 256)	\
+		      / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+#if (CYC_PER_TICK_M & 0xFF) > 0x7F
+#define CYC_PER_TICK ((CYC_PER_TICK_M >> 8) + 1)
+#else
+#define CYC_PER_TICK (CYC_PER_TICK_M >> 8)
+#endif
+
+#endif
+
+#define MAX_TICKS ((COUNTER_MAX / CYC_PER_TICK) - 1)
+#define MAX_CYCLES (MAX_TICKS * CYC_PER_TICK)
+
+/* Minimum cycles in the future to try to program.  Note that this is
+ * NOT simply "enough cycles to get the counter read and reprogrammed
+ * reliably" -- it becomes the minimum value of the LOAD register, and
+ * thus reflects how much time we can reliably see expire between
+ * calls to elapsed() to read the COUNTFLAG bit.  So it needs to be
+ * set to be larger than the maximum time the interrupt might be
+ * masked.  Choosing a fraction of a tick is probably a good enough
+ * default, with an absolute minimum of 1k cyc.
+ */
+#define MIN_DELAY MAX(32, (CYC_PER_TICK/16))
```

Comment:
Hey, I recognize that comment.  I wrote all that for SysTick, and while the architecture here seems similar, you should at least update the comment to reflect the hardware actually in use. :)  You don't have a COUNTFLAG bit, the comment still talks about a 1024 cycle minimum, etc....

Also note that while this shortcut was probably OK for SysTick running at dozens of MHz, on a much slower counter you're losing precision in those 32 cycles that you'd ideally like to keep (c.f. the carefully crafted cycle-optimal change just submitted for nrf5).  Ideally you should be able to read/check/set/check the counter to lose at most one cycle.

Though I do agree that this is harder than a free running counter to get right.